go/service/chat_local.go

```diff
@@ -89,19 +88,101 @@ func (h *chatLocalHandler) aggRateLimits(rlimits []*chat1.RateLimit) (res []chat
 	return res
 }
 
+func (h *chatLocalHandler) aggRateLimits(rlimits []chat1.RateLimit) (res []chat1.RateLimit) {
+	m := make(map[string]chat1.RateLimit)
+	for _, l := range rlimits {
+		m[l.Name] = l
+	}
+	for _, v := range m {
+		res = append(res, v)
+	}
+	return res
+}
+
+func (h *chatLocalHandler) cryptKeysWrapper(ctx context.Context, tlfName string) (tlfID chat1.TLFID, canonicalTlfName string, err error) {
+	resp, err := h.boxer.tlf.CryptKeys(ctx, tlfName)
+	if err != nil {
+		return nil, "", err
+	}
+	tlfIDb := resp.TlfID.ToBytes()
+	if tlfIDb == nil {
+		return nil, "", errors.New("invalid TLF ID acquired")
+	}
+	tlfID = chat1.TLFID(tlfIDb)
+	return tlfID, string(resp.CanonicalName), nil
+}
+
+func (h *chatLocalHandler) getInboxQueryLocalToRemote(ctx context.Context, lquery *chat1.GetInboxLocalQuery) (rquery *chat1.GetInboxQuery, err error) {
+	if lquery == nil {
+		return nil, nil
+	}
+	rquery = &chat1.GetInboxQuery{}
+	if lquery.TlfName != nil {
+		tlfID, _, err := h.cryptKeysWrapper(ctx, *lquery.TlfName)
+		if err != nil {
+			return nil, err
+		}
+		rquery.TlfID = &tlfID
+	}
+	rquery.After = lquery.After
+	rquery.Before = lquery.Before
+	rquery.TlfVisibility = lquery.TlfVisibility
+	rquery.TopicType = lquery.TopicType
+	rquery.UnreadOnly = lquery.UnreadOnly
+	rquery.ReadOnly = lquery.ReadOnly
+	rquery.ConvID = lquery.ConvID
+
+	return rquery, nil
+}
+
 // GetInboxLocal implements keybase.chatLocal.getInboxLocal protocol.
-func (h *chatLocalHandler) GetInboxLocal(ctx context.Context, arg chat1.GetInboxLocalArg) (chat1.GetInboxLocalRes, error) {
+func (h *chatLocalHandler) GetInboxLocal(ctx context.Context, arg chat1.GetInboxLocalArg) (inbox chat1.GetInboxLocalRes, err error) {
 	if err := h.assertLoggedIn(ctx); err != nil {
 		return chat1.GetInboxLocalRes{}, err
 	}
+
+	rquery, err := h.getInboxQueryLocalToRemote(ctx, arg.Query)
+	if err != nil {
+		return chat1.GetInboxLocalRes{}, err
+	}
 	ib, err := h.remoteClient().GetInboxRemote(ctx, chat1.GetInboxRemoteArg{
-		Query:      arg.Query,
+		Query:      rquery,
 		Pagination: arg.Pagination,
 	})
-	return chat1.GetInboxLocalRes{
-		Inbox:      ib.Inbox,
-		RateLimits: h.aggRateLimits([]*chat1.RateLimit{ib.RateLimit}),
-	}, err
+	if err != nil {
+		return chat1.GetInboxLocalRes{}, err
+	}
+	inbox = chat1.GetInboxLocalRes{
+		Pagination: arg.Pagination,
+		RateLimits: h.aggRateLimitsP([]*chat1.RateLimit{ib.RateLimit}),
+	}
+
+	for _, convRemote := range ib.Inbox.Conversations {
```

Comment:
This loop needs to make sure that `ctx` has a `userInfoMapper` in it, that is critical for performance.